5주차 과제1 [레스토랑 조회 시스템 구현하기] 제출

Kwakcena opened the pull request:
다시 해서 제출하다보니 커밋이 너무 많네요... 죄송합니다. 

현재 구현된 부분은 아래와 같습니다.
- api 서버에서 레스토랑 지역 목록 가져오기
- api 서버에서 레스토랑 분류 목록 가져오기
- 클릭시 버튼에 체크 표시
- 클릭한 지역 또는 분류를 region, category 상태값에 담기

그리고 비동기 액션에 대한 테스트도 해 보고 싶어서 아샬님이 참고하라고 하신 redux-mock-store 를 참고하여 해당 비동기 액션이 발생할 경우 어떤 action type이 실행되는지를 테스트 해 봤습니다.

앞으로 해야 할 것은 클릭한 지역, 분류에 따라 레스토랑 목록을 화면에 뿌려주는 api를 실행 할 예정입니다.
reducer에서 if문 줄이기 등 다양한 리팩토링도 필요하지만 일단 오늘은 시간이 없어서 여기까지 하고 제출하겠습니다.

저도 차주부터는 1일 1PR을 목표로 할게요 커밋 너무 많아서 죄송해요 !!!

Kwakcena committed:
Init commit

Kwakcena committed:
test(red)-App: 레스토랑 지역 목록 보기

레스토랑의 지역 목록을 볼 수 있도록 한다.

Kwakcena committed:
test(green)-App: 레스토랑 지역 목록 보기

하드 코딩

Kwakcena committed:
test(red)-RestaurantRegions: 지역 목록 분리

지역 목록을 보여주는 컴포넌트를 분리하기 위해 컴포넌트 테스트를 한다.

Kwakcena committed:
test(green)-RestaurantRegions: 통과

지역 목록 보여주기 통과

Kwakcena committed:
test(refactor)-App: regions 컴포넌트 사용

레스토랑 지역 목록을 보여주는 컴포넌트를 사용한다.

Kwakcena committed:
test(red)-RestaurantRegions: props 추가

레스토랑 지역 목록을 props로 받아서 렌더링 하도록 한다.

Kwakcena committed:
test(green)-RestaurantRegions: 통과

이제 App에서도 통과할 수 있도록 하자.

Kwakcena committed:
test(refactor)-App: fixtures를 이용해 통과시킴

나중에 RestaurantRegionsContainer를 이용하면 fixtures를 이용하던 것을
지워야 함.

Kwakcena committed:
test(red)-RestaurantRegionsContainer: 지역 목록

컨테이너에서 해야 할 일
- 상태를 관리하는 함수와 RestaurantRegions을 이어주기
- 지역 목록 보이기

Kwakcena committed:
test(green)-RestaurantRegionsContainer 지역 목록

컨테이너 내부에서 RestaurantRegions 컴포넌트를 사용하기만 하면 됨.

Kwakcena committed:
가짜 useSelector를 사용해서 테스트를 통과함.

현재 RestaurantRegionsContainer는 useSelector를 사용하도록 구현되어 있지
않다. 그래서 이를 사용하기 위해 테스트에서 먼저 useSelector에 대해
사용하는 코드를 작성하고 가짜를 가져다가 썻다.

내 생각에는 이 테스트가 깨져야 하는데 왜 통과되는것인지 모르겠다.

Kwakcena committed:
test(red)-RestaurantRegionsContainer

테스트에서는 useSelector를 이용해 regions를 가져오고 해당 상태를 화면에
뿌려 보여주길 원하지만, 실제 구현에는 하드코딩된 regions가 들어있어
원하는 텍스트를 찾을 수 없기에 실패가 된다.

따라서 실제 구현에서도 useSelector를 이용해 상태를 가져오도록 한다.

Kwakcena committed:
test(red)-App: 컨테이너 컴포넌트 사용

컨테이너 컴포넌트를 실제로 사용하면 리듀서가 만들어져 있지 않으므로
실패하게 된다. regions라는 값을 꺼내올 수 없기 떄문이다.

Kwakcena committed:
Redux 사용 준비

Kwakcena committed:
test(red)-reducer: 초기 상태 사용

이전 상태가 없으면 초기 상태를 사용한다.

Kwakcena committed:
test(green)-reducer: regions 변수 추가

초기상태에 regions를 추가했다.

Kwakcena committed:
test(refactor)-App: App 테스트에서 useSelector 사용

App 테스트에서도 useSelector를 사용해야 테스트가 통과된다.

Kwakcena committed:
Refactor: fixtures에 initialState 추가

일부 코드 리팩토링

Kwakcena committed:
test(red)-reducer: 지역 정보를 set 한다.

서버로부터 지역 목록을 받아오면 이를 regions 배열에 담는 setRegions
액션에 대한 테스트를 진행한다.

Kwakcena committed:
test(green)-reducer: set regions

레스토랑 지역 목록을 regions에 담기

Kwakcena committed:
Refactor: actions 사용

actions를 사용함.
reducer의 action 기본값을 빈 객체로 넣어 type이 없는 경우 방지

Kwakcena committed:
비동기 action 테스트를 위한 redux-mock-store

비동기 action 테스트를 위해 redux-mock-store를 설치한다.

Kwakcena committed:
test(red)-reducer: 비동기 액션 테스트 regions

지역 목록을 불러오는 비동기 액션을 테스트 한다.

Kwakcena committed:
store에 thunk 등록

비동기 액션을 사용하기 위해 미들웨어를 등록한다.

Kwakcena committed:
test(red)-reducer: 비동기 액션 테스트

fetch가 정의되어 있지 않으므로 mocking 한 다음 비동기 액션으로 인해
불리는 action의 type이 무엇인지 확인한다.

Kwakcena committed:
test(green)-reducer: 비동기 액션 테스트 통과

Kwakcena committed:
services 빼먹은 커밋

Kwakcena committed:
Refactor: App 컴포넌트도 통과할 수 있게 함

App에서 dispatch를 사용하고 있으므로 이를 mock함.

Kwakcena committed:
Refactor: 필요없는 코드 삭제

Kwakcena committed:
test(red)-App: 식당 분류 목록 보여주기 테스트

Kwakcena committed:
test(green)-App: 분류 보이기 통과

Kwakcena committed:
Refactor: 카테고리 컴포넌트 추가

Kwakcena committed:
test(red)-RestaurantCategories

카테고리를 props로 전달하여 전달받은 분류 내용을 보여야 하는데, 하드코딩
되어 있어서 제대로 된 값을 보이지 못한다.

Kwakcena committed:
test(green)-RestaurantCategories

props를 추가하여 map을 이용해 보여준다.

Kwakcena committed:
test(red)-RestaurantCategoriesContainer

현재 App 컴포넌트에서 RestaurantCategories를 그대로 쓰고 있고, props를
전달해 주지 않아서 테스트가 꺠진다.

이제 컨테이너 컴포넌트를 만들어 상태를 불러와 전달해 줄 수 있도록 해야
한다.

Kwakcena committed:
test(green)-RestaurantCategoriesContainer

레스토랑 카테고리 컨테이너 컴포넌트 구현

Kwakcena committed:
test(red)-App

실제로 레스토랑 카테고리 컨테이너 컴포넌트를 적용하더래도 리듀서에
categories 에 대한 정의가 없기 때문에 useSelector로 가져올 수 없다.

Kwakcena committed:
test(green)-App: categories를 불러오기

Kwakcena committed:
Refactor: 초기 상태에 categories 추가

Kwakcena committed:
test(red)-setCategories

카테고리 목록을 categories 배열에 담는 reducer 테스트 진행

Kwakcena committed:
test(green)-setCategories

카테고리 목록 담기 테스트 통과

Kwakcena committed:
test(red)-loadCategories

서버로부터 카테고리 목록을 불러오는 테스트 진행

Kwakcena committed:
test(green)-loadCategories

서버로부터 분류 목록 불러오기 테스트 통과

Kwakcena committed:
Refactor: App에 적용

Kwakcena committed:
test(red)-RestaurantRegions

지역 목록 중 하나를 클릭하면 onClick 이벤트가 호출되는지 테스트 함

Kwakcena committed:
test(green)-onClick 이벤트 호출 테스트 통과

Kwakcena committed:
Refactor: props에 onClick 추가

Kwakcena committed:
test(red)-reducer set region test

지역 중 하나를 선택하면 region에 보관할 수 있도록 reducer를 테스트 한다.

Kwakcena committed:
test(green)-reducer set region test 통과

Kwakcena committed:
Refactor: actions 추가, 초기상태 추가

actions에 setRegion 추가, 초기상태에 region 변수 추가

Kwakcena committed:
test(red)- dispatch 호출 테스트

클릭 시 dispatch가 호출되는가

Kwakcena committed:
test(green)- dispatch 호출 통과

onClick의 전달 값으로 name을 주어서 클릭한 해당 지역을 region에 담도록
했다.

Kwakcena committed:
Refactor: actions 사용

Kwakcena committed:
test(red)- 지역 클릭 시 체크 표시

지역을 클릭하면 체크 표시가 되도록 하는 테스트를 수행한다.

Kwakcena committed:
test(green)- 체크 표시 성공

하나의 값만 체크 표시가 되도록 성공

Kwakcena committed:
Refactor: 재사용 가능하도록 함수로 구성

체크 표시를 보여주는 기능은 분류에서도 해야 하는 작업이므로 함수로
분리한다.

Kwakcena committed:
Refactor: 여러번 누르면 체크 표시가 여러개 나타남

이를 해결하기 위한 리팩토링

Kwakcena committed:
test(red)- 분류 클릭시 category에 담기 테스트

Kwakcena committed:
test(green)- 이벤트 핸들러 호출 테스트 통과

Kwakcena committed:
test(red)-카테고리 컨테이너 dispatch 호출 테스트

카테고리 컨테이너에서 분류를 클릭시 dispatch가 호출되는지 테스트 한다.

Kwakcena committed:
test(red)-reducer 카테고리 선택 시 set

컨테이너 컴포넌트에서 dispatch를 사용하기 전에 reducer에서 테스트를
진행한 뒤 사용한다.

Kwakcena committed:
test(green)- reducer 카테고리 set 테스트 통과

Kwakcena committed:
test(red)-region 자체를 전달하도록 수정한다.

Kwakcena committed:
test(green)- region 자체를 전달

Kwakcena committed:
Refactor: 테스트 수정, 체크 표시 인자전달 수정

Kwakcena committed:
test(red)- 분류 자체가 전달되도록 테스트 수정

Kwakcena committed:
Refactor: 분류 인자전달 수정, 체크 표시 테스트 수정

- 분류 목록을 클릭하면 해당 분류의 아이디만 전달하던 것을 객체로
  전달하게 했다.
- 추후에 (V) 표시를 남기는 함수에 전달할 인자로 프로퍼티를 나누어
  전달하기 위함이다.
- 체크 표시가 나오는지에 대한 테스트도 진행했다.

Kwakcena committed:
Refactor: reducer test refactoring

Kwakcena committed:
test(red): 레스토랑 목록 보여주기

지역과 분류를 선택해서 나오는 레스토랑 목록을 보여준다.

Kwakcena committed:
test(green): 레스토랑 목록 하드코딩으로 통과

Kwakcena committed:
Refactor: 레스토랑 목록 컴포넌트 분류

Kwakcena committed:
test(red): props를 받아서 목록을 뿌린다.

Kwakcena committed:
test(green): props로 레스토랑 목록 받아서 뿌리기

hannut91 commented on src/App.test.jsx:
```diff
@@ -0,0 +1,52 @@
+
+describe('App', () => {
+  const dispatch = jest.fn();
+  beforeEach(() => {
```
위에 빈 줄 하나 추가해서 구분해주면 좋을 것 같아요

그리고 `dispatch.mockClear()`를 넣어주는게 더 안전할 것 같아요.

hannut91 commented on src/RestaurantCategories.test.jsx:
```diff
@@ -0,0 +1,36 @@
+
+import { categories } from '../fixtures/restaurants';
+
+describe('RestaurantCategories', () => {
+  it('분류 목록을 받아서 화면에 보여준다.', () => {
+    const categories = [
+      { id: 1, name: '서양식' },
+      { id: 2, name: '동양식' },
+    ];
+
+    const { getByText } = render((
+      <RestaurantCategories categories={categories} />
+    ));
+
+    categories.forEach(({ name }) => {
+      expect(getByText(name)).toBeInTheDocument();
+    });
+  });
+
+  it('분류 클릭 시 이벤트 함수가 호출된다.', () => {
+    const handleClick = jest.fn();
+    const { getByText } = render((
+      <RestaurantCategories
+        categories={categories}
+        onClick={handleClick}
+      />
+    ));
+    fireEvent.click(getByText('한식'));
+    expect(handleClick).toBeCalled();
+  });
+});
```
중복되는 부분을 제거하고 Given을 외부로 분리해봤습니다.

```js
describe('RestaurantCategories', () => {
  const handleClick = jest.fn();

  const renderRestaurantCategories = () =>
    render((
      <RestaurantCategories
        categories={categories}
        onClick={handleClick}
      />
    ));

  beforeEach(() => {
    handleClick.mockClear();
  });

  it('분류 목록을 받아서 화면에 보여준다.', () => {
    const { getByText } = renderRestaurantCategories();

    categories.forEach(({ name }) => {
      expect(getByText(name)).toBeInTheDocument();
    });
  });

  it('분류 클릭 시 이벤트 함수가 호출된다.', () => {
    const { getByText } = renderRestaurantCategories();

    fireEvent.click(getByText('한식'));
    
    expect(handleClick).toBeCalled();
  });
});
```

hannut91 reviewed (changes requested):
TDD를 하실 때 리팩터링과정을 빠트리지 않고 하면 좋을 것 같아요.

Kwakcena committed:
Refactor: Add mockClear

mockClear를 넣는것이 더 안전하다는 리뷰에 따라 추가함.

mockClear 의 역할
모의 생성자 함수와 메서드 호출 기록을 정리.

Kwakcena committed:
Refactor: RestaurantCategories

중복되는 부분을 제거하고 Given을 외부로 분리

Kwakcena committed:
test(green): App test

restaurants 컴포넌트는 props를 받으므로 fixtures 데이터를 주어 테스트
통과

Kwakcena committed:
test(red): RestaurantsContainer

컨테이너 컴포넌트에 내용이 없으므로 에러

Kwakcena committed:
test(green): 레스토랑 목록 보여주기 테스트

컨테이너 컴포넌트에서 레스토랑 목록을 보여주도록 fixtures 데이터를
이용해 보여주고 있다.

Kwakcena committed:
Refactor: App에서 컨테이너 사용

화면에 Restaurants 를 뿌려주는 역할을 컨테이너 컴포넌트 안에
넣어줬으므로 실제 App 컴포넌트에 이를 적용해준다.

Kwakcena committed:
test(red): 컨테이녀 useSelector 사용

useSelector를 사용하여 레스토랑 목록을 가져오는 테스트 진행

Kwakcena committed:
test(green): 레스토랑 목록 useSelector 테스트

컨테이너에 실제로 useSelector를 적용하여 test가 통과하도록 한다.

Kwakcena committed:
Refactor: fixtures 데이터 사용

테스트에 fixtures 데이터를 사용한다.

Kwakcena committed:
Refactor: restaurants 데이터 불러오기

App 컴포넌트에서 restaurants 데이터도 불러오도록 테스트를 하여
통과시킨다.

Kwakcena committed:
test(red): 레스토랑 목록을 restaurants에 추가하기

서버에 요청한 restaurants 목록을 reducer를 이용해서 변수에 추가하는
테스트를 수행한다.

Kwakcena committed:
test(green): 레스토랑 목록 추가하기 통과

reducer로 레스토랑 목록 추가하기 테스트 통과

Kwakcena committed:
Refactor: 명확한 테스트로 수정하기

추가한 restaurants가 실제로 fixtures의 restaurants와 같은지 테스트를
한다.

Kwakcena committed:
test(red): loadRestaurants 비동기 액션 테스트

레스토랑 목록을 불러오는 비동기 액션을 테스트 한다.

Kwakcena committed:
test(red): 레스토랑 목록 불러오기 테스트

regionName, categoryId가 존재하면 레스토랑 목록을 불러오는 비동기 액션
테스트를 진행한다.

Kwakcena committed:
test(green): 비동기 액션 구현

레스토랑 목록을 불러오는 비동기 액션을 구현한다.

Kwakcena committed:
Refactor: mockStore에서 getState 사용

가짜 mockStore에 등록한 state를 getStore를 이용하면 가져올 수 있다.

Kwakcena committed:
막힌 부분

원래는 App 컴포넌트에서 loadRestaurants 를 이용해서 데이터를 api 서버에
요청하려 했으나, App 컴포넌트에서는 상태를 가져오지 않으므로
loadRestaurants에 인자를 전달하지 못함을 깨닫고

RestaurantsContainer에서 이를 처리하려 했습니다. 근데 regionName과
categoryId가 비어있을때에 대한 처리를 어떻게 테스하고 적용해야 할지
모르겠어요.

일단 RestaurantsContainer에서 상태를 가져오는 것이 맞는지부터
햇갈립니다.

hannut91 commented on src/RestaurantCategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,46 @@
+jest.mock('react-redux');
+
+describe('RestaurantCategoriesContainer', () => {
+  const dispatch = jest.fn();
```
여기도 mockClear를 하면 좋을 것 같네요

hannut91 commented on src/RestaurantRegionsContainer.test.jsx:
```diff
@@ -0,0 +1,46 @@
+jest.mock('react-redux');
+
+describe('RestaurantRegionsContainer', () => {
+  const dispatch = jest.fn();
```
여기도 mockClear를 해주면 좋을 것 같아요

hannut91 commented on src/reducer.js:
```diff
@@ -0,0 +1,60 @@
+  restaurants: [],
+};
+
+function newArray(array, value) {
```
newArray가 오히려 이해하는데 방해랑 하고 있는 것 같아요. 

```js

const check = (value) => ({ name, value }) => 
  value === name
       ? { id, name: `${name.replace(/\(V\)/g, '')}(V)` }
       : { id, name: name.replace(/\(V\)/g, '') }

Ccategories
  .map(check(category.name);

hannut91 commented on src/reducer.test.js:
```diff
@@ -0,0 +1,123 @@
+      const store = mockStore({
+        regions: [],
+      });
+      return store.dispatch(loadRegions())
```
async await을 사용하면 더 간단하게 표현할 수 있을거에요

hannut91 reviewed (approved):
고생하셨어요 ㅎㅎ 맘에 안들어서 다시 하시다니 열정이 대단하시네요 ㅎㅎㅎ 지금도 좋지만 더더더더더 좋게 할 수 있을지 고민하시는 모습이 좋네요
